### PR TITLE
[CTRLRFC-980][Dhyana] Dhyana evolution

### PR DESCRIPTION
**Pull request by goldensgui**
Add Frame rate attribute and implement trigger output

### CODE REVIEW

File: include/DhyanaCamera.h [hunk 1]
```diff
@@ -70,6 +70,40 @@ class LIBDHYANA_API Camera
...
+    {
+      SignalEdgeRising = TUOPT_RISING,
+      SignalEdgeFalling = TUOPT_FAILING,
+    };
+
+    enum TucamGain
+    {
+      GainHDR  = TUGAIN_HDR,
+      GainHigh = TUGAIN_HIGH,
+      GainLow  = TUGAIN_LOW
```
Comment (soleil-ica) [inline include/DhyanaCamera.h]: could you please rename the enum keys by adding a small k eg: 
enum TucamSignalEdge
    {
      kSignalEdgeRising = TUOPT_RISING,
      kSignalEdgeFalling = TUOPT_FAILING,
    };

File: include/DhyanaCamera.h [hunk 2]
```diff
@@ -135,6 +169,14 @@ class LIBDHYANA_API Camera
...
     bool isAcqRunning() const;
 
+    void getFPS(double& fps);	
+    void getTriggerMode(TucamTriggerMode& mode){mode = m_tucam_trigger_mode;};
+    void setTriggerMode(TucamTriggerMode mode){m_tucam_trigger_mode = mode;};
+    void getTriggerEdge(TucamTriggerEdge& edge){edge = m_tucam_trigger_edge_mode;};
+    void setTriggerEdge(TucamTriggerEdge edge){m_tucam_trigger_edge_mode = edge;};
+    void getOutputSignal(int port, TucamSignal& signal, TucamSignalEdge& edge, int& delay, int& width);
+    void setOutputSignal(int port, TucamSignal signal, TucamSignalEdge edge=SignalEdgeRising, int delay=-1, int width=-1);
+
```
Comment (soleil-ica) [inline include/DhyanaCamera.h]: move the functions impl into the .cpp , like for the getFPS()
also add  DEB_MEMBER_FUNCT()

File: include/DhyanaCamera.h [hunk 3]
```diff
@@ -179,7 +221,13 @@ class LIBDHYANA_API Camera
     // Buffer control object
     SoftBufferCtrlObj   m_bufferCtrlObj;
 	CSoftTriggerTimer*	m_internal_trigger_timer;
+    double              m_fps;
 	unsigned short 		m_timer_period_ms;
+    TucamTriggerMode    m_tucam_trigger_mode;
+    TucamTriggerEdge    m_tucam_trigger_edge_mode;
+    TUCAM_TRGOUT_ATTR tgroutAttr1;
+    TUCAM_TRGOUT_ATTR tgroutAttr2;
+    TUCAM_TRGOUT_ATTR tgroutAttr3;
```
Comment (soleil-ica) [inline include/DhyanaCamera.h]: 1/ move this into the //TUCAM stuff above
2/ add m_ prefix to members

File: src/DhyanaCamera.cpp [hunk 4]
```diff
@@ -123,6 +125,24 @@ void Camera::init()
 	
 	//initialize TUCAM Event used when Waiting for Frame
 	m_hThdEvent = NULL;
+
+	tgroutAttr1.nTgrOutPort = 0;
+	tgroutAttr1.nTgrOutMode = TucamSignal::SignalReadEnd;
+	tgroutAttr1.nEdgeMode = 0;
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: Why the init value is 0 and not 1 like the 2 others ?

File: src/DhyanaCamera.cpp [hunk 5]
```diff
@@ -444,8 +476,10 @@ void Camera::AcqThread::threadFunction()
 		
 		Timestamp t1_capture = Timestamp::now();
 		double delta_time_capture = t1_capture - t0_capture;
+
 		DEB_TRACE() << "Capture all frames elapsed time = " << (int) (delta_time_capture * 1000) << " (ms)";				
-		
+		DEB_ALWAYS() << "FPS : " << std::setprecision(2) << m_cam.m_fps;
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: if this is necessary to display it, use DEB_TRACE instead of DEB_ALWAYS

File: src/DhyanaCamera.cpp [hunk 6]
```diff
@@ -1104,6 +1138,139 @@ void Camera::getFirmwareVersion(std::string& version)
...
+    fps = m_fps;
+}
+
 //-----------------------------------------------------
-//
-//-----------------------------------------------------  
\ No newline at end of file
+// Set trigger outpout on selected channel
+//-----------------------------------------------------  
+void Camera::setOutputSignal(int port, TucamSignal signal, TucamSignalEdge edge, int delay, int width)
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: is it "channel" or "port"?
choose the same name

File: src/DhyanaCamera.cpp [hunk 7]
```diff
@@ -1104,6 +1138,139 @@ void Camera::getFirmwareVersion(std::string& version)
...
+
+		tgroutAttr1.nTgrOutMode = tgroutAttr.nTgrOutMode;
+		tgroutAttr1.nEdgeMode = tgroutAttr.nEdgeMode;
+		tgroutAttr1.nDelayTm = tgroutAttr.nDelayTm;
+		tgroutAttr1.nWidth = tgroutAttr.nWidth;
+		break;
+
+	case 1:
+
+		tgroutAttr2.nTgrOutMode = signal;
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: this line is doubled

File: src/DhyanaCamera.cpp [hunk 8]
```diff
@@ -1104,6 +1138,139 @@ void Camera::getFirmwareVersion(std::string& version)
...
+		}
+		tgroutAttr2.nTgrOutMode = tgroutAttr.nTgrOutMode;
+		tgroutAttr2.nEdgeMode = tgroutAttr.nEdgeMode;
+		tgroutAttr2.nDelayTm = tgroutAttr.nDelayTm;
+		tgroutAttr2.nWidth = tgroutAttr.nWidth;
+		break;
+
+	case 2:
+
+		tgroutAttr3.nTgrOutMode = signal;
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: idem

File: src/DhyanaCamera.cpp [hunk 9]
```diff
@@ -1104,6 +1138,139 @@ void Camera::getFirmwareVersion(std::string& version)
...
+		break;
+
+	default:
+		THROW_HW_ERROR(Error) << "Unable to set Output signal port " << port;
+		break;
+	}
+}
+
+//-----------------------------------------------------
+// Get trigger outpout on selected channel
```
Comment (soleil-ica) [inline src/DhyanaCamera.cpp]: idem for channel vs port

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**